// dena-client/op-client/lib/op.js
/**
 * Created by yong.liu on 2015/5/7.
 */

/**
 * op client
 */

var zlib = require('zlib');
var request = require('request');
var querystring = require('querystring');

var Op = function () {
    this.fn = null;
    this.actions = [];
    this.customValue = {};
    this.options = {}
};

module.exports = Op;

Op.prototype.run = function () {
    var self = this;

    var timer = setInterval(breath, self.breathIntervalTime);

    function breath() {
        if (self.fn === null) {
            self.fn = self.actions.shift();
            return
        }

        if (self.fn() === undefined) {
            if (self.actions.length) {
                self.fn = self.actions.shift();
            } else {
                console.log('clear timer')
                clearInterval(timer);
            }
        }
    }
};

Op.prototype.getThinkTime = function () {
    return Math.ceil(Math.random() * this.thinkTimeMax / 1000);
}

Op.prototype.init = function (params, cb) {
    this.port = params.port;
    this.host = params.host;

    this.options = {
        agent: false,
        headers: {
            'Accept-Encoding': 'gzip'
        }
    };

    setImmediate(cb);
}

Op.prototype.request = function (path, param, callback) {
    this.options.url = 'http://' + this.host + ':' + this.port + path;
    this.options.body = querystring.stringify(param);

    var req = request.post(this.options);

    req.on('response', function (res) {
        var chunks = [];

        res.on('data', function (chunk) {
            chunks.push(chunk);
        });

        res.on('end', function () {
            var message = {};
            message.code = res.statusCode;

            var buffer = Buffer.concat(chunks);
            var encoding = res.headers['content-encoding'];

            switch (encoding) {
                case 'gzip':
                    zlib.gunzip(buffer, function (err, decoded) {
                        message.body = decoded && decoded.toString()
                        callback(message);
                    });

                    break;
                case 'deflate':
                    zlib.inflate(buffer, function (err, decoded) {
                        message.body = decoded && decoded.toString()
                        callback(message);
                    });

                    break;
                default:
                    message.body = buffer.toString()
                    callback(message);
            }
        });
    });

    req.on('error', function (err) {
        callback(err);
    });
};
